src/app/user.ts

```diff
@@ -0,0 +1,42 @@
+import {UserData} from './user-data';
+export class User {
```

Comment:
un appunto. le _interface_ di typescript non sono esattamente come le interfacce in java. in java un interfaccia specifica behaviour, in typescript invece anche gli oggetti letterali possono essere un "interfaccia". io questa classe l'avrei fatta così:
```ts
interface UserData {
//...
}

interface User {
  readonly id: number;
  readonly mail: string;
  readonly password: string;
  readonly userData: UserData;
}
```
in questo modo c'è un match preciso con la struttura della risorsa. inoltre in generale classi readonly sono più veloci rispetto a quelle con i setter. quando devi modificare un elemento si crea un nuovo oggetto pescando da quello vecchio le proprietà rimaste uguali. non sono sicurissimo della performance ma uno dei vantaggi più evidenti è che il codice diventa automaticamente parallelizzabile perché non modifichi mai nulla e non hai bisogno di mutex

Reply:
e possiamo anche mescolare campi readonly e non?

Reply:
l'ideale sarebbe avere solo interfacce puramente readonly però se proprio vuoi avere campi modificabili fallo pure sporadicamente.

come workflow generale io consiglio sempre di mettere tutto readonly/const/final e poi toglierlo dove serve